client-js/src/client/errors.js

```diff
@@ -0,0 +1,203 @@
+
+  /**
+   * @param {!Error} error the error caught from {@code fetch} invocation
+   */
+  constructor(error) {
+    super(error.message, error);
+  }
+}
+
+/**
+ * An abstract error indicating an invalid server behaviour.
+ *
+ * @extends SpineWebError
+ * @abstract
+ */
+export class ServerError extends SpineWebError {
+}
+
+/**
+ * An abstract error indicating an invalid client behaviour.
+ *
+ * @extends SpineWebError
+ * @abstract
+ */
+export class ClientError extends SpineWebError {
+}
+
+/**
+ * An error which occurs when sending off a request to Spine server endpoint results
+ * with a response with `5xx` status code.
+ *
+ * Indicates an unhandled exception was thrown upon the request processing.
+ *
+ * @extends ServerError
+ */
+export class InternalServerError extends ServerError {
+
+  /**
+   * @param {!Response} response the server response caused this error
+   */
+  constructor(response) {
+    super(response.statusText, response);
+  }
+}
+
+/**
+ * An error which occurs when sending off a request to Spine server endpoint results with a
+ * malformed response which can't be processed.
+ *
+ * @extends ServerError
+ */
+export class ResponseProcessingError extends ServerError {
+
+  /**
+   * @param {?Error} cause the reason why this error occurred
+   */
+  constructor(cause) {
+    super(cause.message || cause, cause);
+  }
+}
+
+/**
+ * An error which occurs when sending off a request to Spine server endpoint results
+ * with a response with `4xx` status code.
+ *
+ * @extends ClientError
+ */
+export class RequestProcessingError extends ClientError {
+
+  /**
+   * @param {!Response} response the server response caused this error
+   */
+  constructor(response) {
+    super(response.statusText, response);
+  }
+}
+
+/**
+ * An error which occurs when sending off a command to Spine server endpoint results with a response
+ * containing {@code spine.base.Ack} with error status. It means that a command message was rejected further
+ * processing (e.g. because of a validation error).
+ *
+ * @extends ClientError
+ */
+export class CommandProcessingError extends ClientError {
+  /**
+   * @param {spine.base.Error} error the technical error occurred upon receiving the request and
+   *                                  no further processing would occur.
+   */
+  constructor(error) {
+    super(error.message, error);
+  }
+
+  /**
+   * Returns the type of {@code spine.base.Error}.
+   *
+   * @return {string}
+   */
+  type() {
+    return this.getCause().type;
+  }
+
+  /**
+   * Returns the code of {@code spine.base.Error}.
+   *
+   * @return {number}
+   */
+  code() {
+    return this.getCause().code;
+  }
+
+  /**
+   * Returns an optional validation error object error indicating that a
+   * message sent to the server did not pass validation.
+   *
+   * @return {?spine.validate.ValidationError}
+   */
+  validationError() {
+    return this.getCause().validationError;
```

Comment:
Not every command handling error may be a validation error. Maybe validation error should inherit the CommandHandlingError.

Reply:
If there is an error during command handling it can be both a generic client error, an ack error (which can have a validation error or some other error inside of it), and a server error. All of them are basically *command handling errors*. So it seems like any of those can be a _cause_ for a CommandHandlingError. 

If we need a check as for whether the command is guaranteed not to be applied by the server, this way we can add a boolean method to the error. `CommandHandlingError#assureCommandUnhandled()` returning a value base on its cause.